src/cloudai/report_generator/training/parser.py

```
@@ -141,9 +164,14 @@ def _group_by_step(scalars: list[Scalar]) -> dict[int, list[Scalar]]:
         return dict(sorted(by_step.items()))
 
     @staticmethod
-    def _get_from_dict(raw: dict, path: str) -> Any:
-        """Read a dotted path (e.g. 'model.num_layers') from a nested dict; None if any key is missing."""
-        return reduce(lambda value, key: value.get(key) if isinstance(value, dict) else None, path.split("."), raw)
+    def _get_from_dict(source: dict, path: str) -> Any:
```

Ruff (0.15.20)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `_get_from_dict`

(ANN401)